mitxgraders/helpers/calc.py

```diff
@@ -447,20 +448,34 @@ def parse_algebra(self):
         # Predefine recursive variables.
         expr = Forward()
 
-        # Handle variables passed in. They must start with a letter
-        # and may contain numbers and underscores afterward.
-        inner_varname = Combine(Word(alphas, alphanums + "_") + ZeroOrMore("'"))
-        # Alternative variable name in tensor format
-        # Tensor name must start with a letter, continue with alphanums
-        # Indices may be alphanumeric
-        # e.g., U_{ijk}^{123}
-        upper_indices = Literal("^{") + Word(alphanums) + Literal("}")
+        # Handle variables passed in. Variables should be of the form
+        #   front + subscripts + lower_indices + upper_indices + tail
+        # where:
+        #   front (required):
+        #       starts with alpha, followed by alphanumeric
+        #   subscripts (optional):
+        #       any combination of alphanumeric and underscores
+        #   lower_indices (optional):
+        #       like subscripts, but with curly braces, e.g., U_{ijk}. Intended
+        #       for consistency with upper-index tensor notation.
+        #   upper_indices (optional):
+        #       e.g., U_{ijk}^{123}
+        #   tail:
+        #       any number of primes
+        front = Word(alphas, alphanums)
+        # ~FollowedBy prevents capturing lower_indices starting character.
+        subscripts = Word(alphanums + '_') + ~FollowedBy('{')
         lower_indices = Literal("_{") + Word(alphanums) + Literal("}")
-        tensor = Combine(Word(alphas, alphanums) + Optional(lower_indices) +
-                         Optional(upper_indices) + ZeroOrMore("'"))
-        # Test for tensor first, then generic variable with underscores for backwards
-        # compatability
-        varname = Group(tensor | inner_varname)("variable")
+        upper_indices = Literal("^{") + Word(alphanums) + Literal("}")
+        tail = ZeroOrMore("'")
+        inner_varname = Combine(
+            front +
+            Optional(subscripts) +
+            Optional(lower_indices) +
+            Optional(upper_indices) +
+            Optional(tail)
```

Comment:
I think this doesn't need to be optional on the tail?

Reply:
True. It's effectively optional already by `ZeroOrMore`. I'll switch it to a comment.